internal/telemetry/telemetry.go

```diff
@@ -0,0 +1,180 @@
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"contrib.go.opencensus.io/exporter/ocagent"
+	datadog "github.com/DataDog/opencensus-go-exporter-datadog"
+	"github.com/hashicorp/go-hclog"
+	octrace "go.opencensus.io/trace"
+	"go.opencensus.io/zpages"
+)
+
+type Telemetry struct {
+	config struct {
+		enableOpenCensusExporter  bool
+		openCensusExporterOptions []ocagent.ExporterOption
+
+		enableDatadogExporter  bool
+		datadogExporterOptions datadog.Options
+
+		enableZpages bool
+		// Address to listen on for zpages. Defaults to 127.0.0.1:9999
+		zpagesAddr string
+	}
+
+	// Logger is the logger to use. This will default to hclog.L() if not set.
+	log hclog.Logger
+
+	// Configured and running exporters that need to be closed
+	zpagesServer *http.Server
+
+	// Configured exporters that need to be registered and closed
+	exporters []exporter
+}
+
+// Exporter is an OpenCensus exporter
+type exporter interface {
+	// Instructs the exporter to register itself (for traces, views, or both)
+	register()
+
+	// Instructs the exporter to flush data and shut down.
+	close() error
+}
+
+// NewTelemetry initializes the telemetry components.
+func NewTelemetry(opts ...Option) (Telemetry, error) {
+	var t Telemetry
+	for _, opt := range opts {
+		opt(&t)
+	}
+
+	if t.log == nil {
+		t.log = hclog.L().Named("telemetry")
+	}
+	log := t.log
+
+	config := t.config
+
+	if config.enableOpenCensusExporter {
+		log.Debug("Creating the OpenCensus agent exporter")
+
+		e, err := newOpenCensusAgentExporter(config.openCensusExporterOptions, log)
+		if err != nil {
+			return t, err
+		}
+
+		t.exporters = append(t.exporters, e)
+	}
+
+	if config.enableDatadogExporter {
+		log.Debug("Starting the Datadog exporter")
+
+		e, err := newDatadogExporter(config.datadogExporterOptions, log)
+		if err != nil {
+			return t, err
+		}
+
+		t.exporters = append(t.exporters, e)
+	}
+
+	// Run zPages
+	if config.enableZpages {
+		zPagesMux := http.NewServeMux()
+		zpages.Handle(zPagesMux, "/debug")
+
+		var zpagesAddr string
+		if config.zpagesAddr == "" {
+			zpagesAddr = "127.0.0.1:9999"
+		} else {
+			zpagesAddr = config.zpagesAddr
+		}
+
+		srv := http.Server{
+			Addr:    zpagesAddr,
+			Handler: zPagesMux,
+		}
+
+		t.zpagesServer = &srv
+	}
```

Comment:
Is there ever a chance where the config is all three of these? Would it be better to write these setup if-statements as a switch on the config type? 🤔

Reply:
Yup, there could be more than one! It's not uncommon to have jeager for traces and ddog for metrics, for example, so you might want both.